crates/rnote-engine/src/store/render_comp.rs

```diff
@@ -242,28 +245,39 @@ impl StrokeStore {
         viewport: Aabb,
         image_scale: f64,
     ) {
-        let keys = self.render_components.keys().collect::<Vec<StrokeKey>>();
-
-        for key in keys {
+        // use the rtree to reduce the number of keys to search through
+        // for now we are using directly the tree because we want to iter without actually
+        // collecting elements
+        let viewport_extended =
+            viewport.extend_by(viewport.extents() * render::VIEWPORT_EXTENTS_MARGIN_FACTOR);
+
+        // we want to iterate on the keys that are in the viewport using the
+        // rtree but also get from this the keys that are not in here
+        // for that also create a slotmap of keys that are in the viewport
+        // so that we can iterate a second time on keys and filter on elements not in the slotmap
+        let mut hashmap_in_viewport: HashMap<StrokeKey, ()> = HashMap::new();
```

Comment:
instead of collecting the keys as a side-effect of the `.map()` adapter, consider collecting into a `HashSet`. This way we can still only iterate once over all keys of `render_components` and check whether the key is in this hashset before calculating the bounds.